Cr 1022 updates to individual request

Paul-Joel opened the pull request:
# Motivation and Context
Updates to improve text and sign posting of individual journeys

# What has changed

- Added function to call campaign site links in centrally managed way

- Removed generic Request Code Intro page functionality and replaced with specific Request Individual code intro page as it is the only one that will be used

- Added post method to Individual intro page to allow for session detection

- Various template updates to add trailing removal markers and collapse error messaging

- Added new 'I don't have a fixed address' collapsible

- Updated expired template to update in page link

- Removed 'new' from individual request text throughout

- Added panel to individual postal journey to declare what the envelope will look like

- Updated so that if user selects to request an individual form at the select method point in the household journey they are presented with the Individual intro screen before continuing, but as an individual.

- Updated unit tests, including adding step to cope with post on individual intro throughout individual tests

# Cucumber update now required for updated error message
https://github.com/ONSdigital/census-rh-cucumber/pull/97

# How to test?
Follow individual UAC request path from /start/, see revised intro apge and marking throughout
Start a request household code, but select to request an individual on the select method of delivery page, and switch to the individual journey without selecting the address again

# Links
https://collaborate2.ons.gov.uk/jira/browse/CR-1022
https://github.com/ONSdigital/census-rh-cucumber/pull/97

# Screenshots (if appropriate):

AmeeshKutthan commented:
Will review as a priority

AmeeshKutthan commented:
Code review comments below:

1. The successful 'customer/user' journey for the AC fulfilment for both SMS and Post work correctly

2. There are a number of issues relating to the Error pages. They have been detailed below:

- These web pages/screens contain incorrect copy (i.e. text content) for the error messages

- These web page/screens layouts are missing some elements relating to the error

- The error messages seem erroneous and emanate incorrect behaviour as per the data entry by the user

All associated screenshots have been attached to this code review comment with comments within as a part of their filename.

3. Even though there is a flow diagram associated with the Jira issue, relating to the Paper fulfilment, there are no code changes within this code commit referring to the Paper fulfilment.
I believe that the Paper fulfilment implementation is detailed within CR-1142. If so, then maybe the Paper fulfilment flow diagram should be removed from this CR, i.e. CR-1022 and added to CR-1142 instead.

<img width="1792" alt="Screenshot 2020-10-20 at 14 52 47_text_incorrect" src="https://user-images.githubusercontent.com/72972620/96616328-59a45d00-12fa-11eb-9ed2-c95ee66d88b9.png">
<img width="1792" alt="Screenshot 2020-10-20 at 14 57 18_text_incorrect" src="https://user-images.githubusercontent.com/72972620/96616340-5dd07a80-12fa-11eb-9315-a2e23416f89f.png">
<img width="1792" alt="Screenshot 2020-10-20 at 15 01 03_text_and_layout_incorrect" src="https://user-images.githubusercontent.com/72972620/96616344-5e691100-12fa-11eb-991d-cf6e3fffd298.png">
<img width="1792" alt="Screenshot 2020-10-20 at 15 04 20_seems_correct_but_layout_is_different_in_flow_diagram" src="https://user-images.githubusercontent.com/72972620/96616348-5f01a780-12fa-11eb-9ea7-18eef47b8b8b.png">
<img width="1792" alt="Screenshot 2020-10-20 at 15 05 09_text_incorrect" src="https://user-images.githubusercontent.com/72972620/96616352-6032d480-12fa-11eb-8cad-6f415536a2cb.png">
<img width="1792" alt="Screenshot 2020-10-20 at 15 06 44_text_and_layout_incorrect_missing_elements" src="https://user-images.githubusercontent.com/72972620/96616353-60cb6b00-12fa-11eb-82de-9f943f21d08e.png">
<img width="1792" alt="Screenshot 2020-10-20 at 15 08 35_text_incorrect" src="https://user-images.githubusercontent.com/72972620/96616355-61640180-12fa-11eb-9121-1a5f8dd22603.png">
<img width="1792" alt="Screenshot 2020-10-20 at 15 09 18_text_incorrect" src="https://user-images.githubusercontent.com/72972620/96616359-62952e80-12fa-11eb-95d2-91fd15c836dc.png">
<img width="1792" alt="Screenshot 2020-10-20 at 15 11 06_text_incorrect_and_no_ui_in_flow_diagram" src="https://user-images.githubusercontent.com/72972620/96616364-632dc500-12fa-11eb-8255-d158ab8a599d.png">
<img width="1792" alt="Screenshot 2020-10-20 at 15 49 02_this_error_occurred_when_I_underwent_an_SMS_flow_then_went_back_and_commenced_a_POST_flow" src="https://user-images.githubusercontent.com/72972620/96616369-645ef200-12fa-11eb-967c-753ee3962059.png">

AmeeshKutthan reviewed (commented):
I have added a code review comment to the bottom of this Github pull request.
All relevant screenshot images have also been attached to this comment.

AmeeshKutthan commented:
Thanks. On my radar. Will jump back onto this once CR-468 has been completed

AmeeshKutthan reviewed (commented):
Code Review:

I have now re-reviewed this CR. Once again, utilising the following reference design document:

https://overflow.io/s/4MYGYYZ1?node=6aaaa044

The following issues have been identified:

1. The word 'new' is missing on this page and the sub-text should not be in bold:

<img width="1792" alt="Screenshot 2020-10-26 at 15 41 06_missing_new_word_text_and_sub_text_is_bold" src="https://user-images.githubusercontent.com/72972620/97201157-09ba1000-17aa-11eb-817f-d1c599afc1cf.png">

2. The word 'new' is missing on this page and the sub-text should not be in bold:

<img width="1792" alt="Screenshot 2020-10-26 at 15 42 19_missing_new_word_and_sub_text_bold" src="https://user-images.githubusercontent.com/72972620/97201325-39691800-17aa-11eb-9683-e4a1d35e4bb6.png">

3. 'Enter your mobile' should state 'Select an answer', per error message convention across site, and sub text error message should not be in bold.
Please note, this error screen is not detailed within the above design document

<img width="1792" alt="Screenshot 2020-10-26 at 15 43 51_text_incorrect_sub_text_is_bold" src="https://user-images.githubusercontent.com/72972620/97201761-bdbb9b00-17aa-11eb-8cee-b674f721c4a2.png">

4. 'You have not entered a postcode' should state 'Select an answer', as per error message convention across the current site.
Please note, this error screen is not detailed within the above design document:

<img width="1792" alt="Screenshot 2020-10-26 at 15 46 26_incorrect_text" src="https://user-images.githubusercontent.com/72972620/97202181-30c51180-17ab-11eb-92ce-dbfc420e0fb3.png">

5. 'Select your address' instead of 'What is your address?' is incorrectly detailed, and the bold sub text error message has been added, even though it is not detailed within the design document. An additional message 'To send a paper questionnaire.....' has been omitted, also:

<img width="1792" alt="Screenshot 2020-10-26 at 15 48 09_sub_text_added_in_bold_and_incorrect_message_text_and_omitted_text" src="https://user-images.githubusercontent.com/72972620/97202940-2b1bfb80-17ac-11eb-97f2-5c6193d4a574.png">

6. The word 'new' is missing from the message and the sub text is incorrectly in a bold font:

<img width="1792" alt="Screenshot 2020-10-26 at 15 50 05_bold_text_being_used_missing_new_word" src="https://user-images.githubusercontent.com/72972620/97203197-6ae2e300-17ac-11eb-8f81-a2bc9cc84095.png">

7. The word 'new' is missing from the message and the sub text is incorrectly in a bold font:

<img width="1792" alt="Screenshot 2020-10-26 at 15 50 59_new_word_missing_and_use_of_bold_text" src="https://user-images.githubusercontent.com/72972620/97203352-96fe6400-17ac-11eb-83c7-d9283fe3867c.png">

8. No sub-text is detailed stating 'This is who the letter will be addressed to':

<img width="1792" alt="Screenshot 2020-10-26 at 15 51 37_no_sub_text" src="https://user-images.githubusercontent.com/72972620/97203470-c1502180-17ac-11eb-9d72-19d927007121.png">

9. No sub-text is detailed stating 'This is who the letter will be addressed to':

<img width="1792" alt="Screenshot 2020-10-26 at 15 52 18_no_sub_text" src="https://user-images.githubusercontent.com/72972620/97203520-d75de200-17ac-11eb-981e-8f0c18c65f12.png">

10. The word 'new' is missing within the sub-text copy:

<img width="1792" alt="Screenshot 2020-10-26 at 15 53 39_missing_new_word" src="https://user-images.githubusercontent.com/72972620/97203644-04aa9000-17ad-11eb-8503-4a048660328d.png">

<img width="1792" alt="Screenshot 2020-10-26 at 17 03 55_missing_new_word" src="https://user-images.githubusercontent.com/72972620/97203973-6b2fae00-17ad-11eb-89bd-06fc8c50d1bc.png">

Paul-Joel commented:
Response to latest comments

1. (and throughout) - have checked with design re: the overflow inconsistency in the journey of ‘an individual access code’ or ‘a new individual access code’ - we have agreed that it should be as implemented, and that 'new' is not required. Also the sub-text should be bold as this is set automatically for declared 'legends' on forms in the design system (DS). The overflow DS version is out of date.
2. As 1.
3. The error should not be 'Select an answer' as there are no answers to select from. Required error is detailed in board 3 HH AC Fulfilment in the Overflow - it is not detailed in the individual because it does not change between the household and individual.
4. Also should not be 'Select an answer', as there are no answers to select from
5. As detailed before, the current solution does not implement the address typeahead functionality - the presentation is correct. As this is not the paper questionnaire journey, it should not be showing paper questionnaire text - this is an error in the Overflow.
6. As 1.
7. As 1.
8. Fixed
9. Fixed
10. As 1.
11. Back links in RHUI have not yet been implemented - their use is still under discussion

AmeeshKutthan reviewed (approved):
All the required changes have been implemented correctly.
Screenshots detailed below:

<img width="1792" alt="Screenshot 2020-10-30 at 13 20 26" src="https://user-images.githubusercontent.com/72972620/97711040-78ea7980-1ab4-11eb-898d-676154796051.png">

<img width="1792" alt="Screenshot 2020-10-30 at 13 20 41" src="https://user-images.githubusercontent.com/72972620/97711050-7d169700-1ab4-11eb-8fc0-caf525171913.png">

<img width="1792" alt="Screenshot 2020-10-30 at 13 22 44" src="https://user-images.githubusercontent.com/72972620/97711054-7daf2d80-1ab4-11eb-920b-7804737bd278.png">

<img width="1792" alt="Screenshot 2020-10-30 at 13 23 07" src="https://user-images.githubusercontent.com/72972620/97711060-7ee05a80-1ab4-11eb-9228-c6c2c45b84dc.png">

MartinHumphrey commented:
I do not know the motivation for treating individual requests and household requests for a new UAC differently but it has undesirable consequences in the following scenario:
- Enter a valid UAC, on the RH /start screen. The uprn and postal address details will be saved to the session. The caseType will not be saved to the session although available.
- Use the browser back arrow to go to the RH /start screen. Note you have **not confirmed** your address.
- If you request a new household access code you will be asked for your postcode, to select and confirm your address, and be  able to request a Text or Post message successfully. 
- If you request a new individual UAC, as you have a session you are asked to select a Text or Post message without entering a postcode etc. You have **not confirmed** the address associated with the UAC you entered.
- Whether you select to send a Post or Text message the application will later fail as it attempts to use caseType from the session but it is not available.
It seems the idea is if you have a session it contains the information necessary to action a fulfilment request, but household and individual requests are treated differently (?) and in fact case type is not available.

Paul-Joel commented:
Have updated checks to ensure that household>individual only jumps to select delivery method if address was confirmed

MartinHumphrey commented on app/utils.py:
```diff
@@ -192,26 +219,39 @@ def validate_uk_mobile_phone_number(number, locale):
             if locale == 'cy':
                 # TODO: Add Welsh Translation
-                raise InvalidDataErrorWelsh('The mobile phone number is not a UK mobile number')
+                raise InvalidDataErrorWelsh('Enter a UK mobile number in a valid format, for example, '
```
Not very DRY. For repetitively used string literals or constant we could declare Them as a variable at the start of the module or even import them from __init__.py?
